src/Components/Server/src/Circuits/RemoteRenderer.cs

```diff
@@ -116,79 +123,142 @@ protected override Task UpdateDisplayAsync(in RenderBatch batch)
+                Log.BeginUpdateDisplayAsync(_logger, _client.ConnectionId);
+                await _client.SendAsync("JS.RenderBatch", Id, pending.BatchId, pending.Data);
             }
-            catch (Exception syncException)
+            catch (Exception e)
             {
-                pendingRenderInfo.TrySetException(syncException);
+                Log.SendBatchDataFailed(_logger, e);
             }
 
-            // When the render is completed (success, fail, or timeout), stop tracking it
-            return pendingRenderInfo.Task.ContinueWith(task =>
-            {
-                _pendingRenders.TryRemove(renderId, out var ignored);
-                if (task.IsFaulted)
-                {
-                    UnhandledException?.Invoke(this, task.Exception);
-                }
-            });
+            // We don't have to remove the entry from the list of pending batches if we fail to send it or the client fails to
+            // acknowledge that it received it. We simply keep it in the queue until we receive another ack from the client for
+            // a later batch (clientBatchId > thisBatchId) or the circuit becomes disconnected and we ultimately get evicted and
+            // disposed.
         }
 
-        public void OnRenderCompleted(long renderId, string errorMessageOrNull)
+        public void OnRenderCompleted(long incomingBatchId, string errorMessageOrNull)
         {
-            if (_pendingRenders.TryGetValue(renderId, out var pendingRenderInfo))
+            if (_disposing)
+            {
+                // Disposing so don't do work.
+                return;
+            }
+
+            // We peek first in case we are receiving a duplicate or out of order ack from
+            // the client.
+            while (PendingRenderBatches.TryPeek(out var entry))
             {
-                if (errorMessageOrNull == null)
+                // Dequeue entries until we sync with the render batch number received.
+                if (incomingBatchId < entry.BatchId)
                 {
-                    pendingRenderInfo.TrySetResult(null);
+                    _logger.LogInformation($"Incoming batch {incomingBatchId} already acknowledged.");
+                    // We successfully received an ack for a pending batch left.
+                    // Which we already processed so no work.
+                    break;
                 }
-                else
+
+                // batchId >= entry.BatchId -> We are in sync or we missed an ack from the client.
+                // We simply catch up as a bigger number means the client already rendered
+                // previous batches successfully.
+                if (incomingBatchId >= entry.BatchId)
                 {
-                    pendingRenderInfo.TrySetException(
-                        new RemoteRendererException(errorMessageOrNull));
+                    _logger.LogInformation($"Acknowledging batch {entry.BatchId} ack sequence {incomingBatchId}.");
```

Comment:
These should be Trace or Debug. It's not very valuable to the user as Info messages.

Reply:
Needs to use LoggerMessage

Reply:
> Needs to use LoggerMessage

Not sure what this means

Reply:
Updated the level to debug